library/core/src/str/mod.rs

```diff
@@ -1231,6 +1231,17 @@ impl<'a, P: Pattern<'a>> SplitInternal<'a, P> {
             },
         }
     }
+
+    #[inline]
+    fn as_str(&self) -> &'a str {
+        // `Self::get_end` doesn't change `self.start`
+        if self.finished {
+            return "";
```

Comment:
This will cause surprising result for those who want to calculate ptr offsets of this &str from the whole &str

Reply:
Is there a use case when you need to do those kinds of tricks?

Reply:
I mean the fix is trivial, but for me, it's not clear if we should provide such guarantees.

Reply:
I would like the return to be a slice of the provided input.

When parsing a http header, I find myself wanting to see how much is remaining in the buffer to determine if I have all the contents (specified by Content-Length).

Example

```
let http_response_buffer: &str;
let mut header_lines = http_response_buffer.split("\r\n");

let mut content_length = None;
/* read through header and set content_length, header ends on empty header_line */
/* ... */

let unread_buffer_size = unsafe {
   header_lines.as_str().as_ptr().offset_from(response_buffer.as_ptr())
};

if content_length.unwrap() < unread_buffer_size {
   panic!("http_response_buffer does not contain response body");
}
```

Reply:
I was thinking about this today and thought that there should be a function (probably in a crate) like
```rust
fn substr_bounds(string: &str, sub: &str) -> Option<Range<usize>> { ... }

let string = "aabb";
let sub = &r[1..3];
assert_eq!(substr_bound(string, sub), Some(1..3));
assert_eq!(substr_bound(string, "ab"), None);
```

It feels natural, but are there any cases when it's useful? Anyway, as said before, for these kinds of things returning an empty substr may be a good idea.

If there is code that may benefit from this, I could make a PR with a fix.

Reply:
It should be related to the semantics of raw pointer comparison.
https://github.com/rust-lang/unsafe-code-guidelines/issues/239